### ISSUE TITLE
Daemonset controller occasionally fails to schedule desired number of pods 

### ISSUE THREAD

**Issue by Joseph-Irving** [kubernetes/kubernetes#75836]
**What happened**:
Occasionally we get a Daemonset stuck with less Current replicas than Desired, causing a Daemonset sync tends to fix it, e.g (add a new node). This happens to all our Daemonsets. I'm using fluentd as an example.
```
NAME      DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
fluentd   59        58        58      58           58          <none>          717d
```
There are no Daemonsets pods in pending nor are there any errors when describing the Daemonset
```
Name:           fluentd
Selector:       app=fluentd
Node-Selector:  <none>
Labels:         app=fluentd
Annotations:    deprecated.daemonset.template.generation: 20
                kubectl.kubernetes.io/last-applied-configuration:
                  {"apiVersion":"extensions/v1beta1","kind":"DaemonSet","metadata":{"annotations":{},"name":"fluentd","namespace":"kube-system"},"spec":{"te...
Desired Number of Nodes Scheduled: 59
Current Number of Nodes Scheduled: 58
Number of Nodes Scheduled with Up-to-date Pods: 58
Number of Nodes Scheduled with Available Pods: 58
Number of Nodes Misscheduled: 0
Pods Status:  58 Running / 0 Waiting / 0 Succeeded / 0 Failed
Pod Template:
  Labels:       app=fluentd
  Annotations:  iam.amazonaws.com/role: cloud_fluentd_logging
                prometheus.io/port: 24231
                prometheus.io/scrape: true
  Containers:
   fluentd:
    Image:      quay.io/uswitch/fluentd:2.1.0-fluentd-1.2-alpine
    Port:       <none>
    Host Port:  <none>
    Limits:
      memory:  4Gi
    Requests:
      cpu:      50m
      memory:   512Mi
    Readiness:  http-get http://:24231/metrics delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:
      FLUENT_UID:   0
      FLUENTD_OPT:  --no-supervisor
    Mounts:
      /fluentd/etc from fluentd-config (ro)
      /var/lib/docker/containers from docker-containers (ro)
      /var/log from logs (rw)
      /var/run/docker.sock from docker (rw)
  Volumes:
   logs:
    Type:          HostPath (bare host directory volume)
    Path:          /var/log
    HostPathType:
   docker-containers:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/docker/containers
    HostPathType:
   docker:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/docker.sock
    HostPathType:
   fluentd-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      fluentd-config
    Optional:  false
Events:
  Type    Reason            Age   From                  Message
  ----    ------            ----  ----                  -------
  Normal  SuccessfulDelete  59m   daemonset-controller  Deleted pod: fluentd-gt9ml
  Normal  SuccessfulCreate  57m   daemonset-controller  Created pod: fluentd-gdr95
  Normal  SuccessfulDelete  54m   daemonset-controller  Deleted pod: fluentd-4xrth
  Normal  SuccessfulCreate  53m   daemonset-controller  Created pod: fluentd-znktv
  Normal  SuccessfulDelete  47m   daemonset-controller  Deleted pod: fluentd-8t6kz
  Normal  SuccessfulCreate  34m   daemonset-controller  Created pod: fluentd-xcvj7
  Normal  SuccessfulDelete  32m   daemonset-controller  Deleted pod: fluentd-gv467
```
**What you expected to happen**:
The Daemonset Pod to be scheduled

**How to reproduce it (as minimally and precisely as possible)**:
Reproducing this state is very tricky, I've had the most success by terminating nodes in our ASG causing a new one to come up.

However I think I've identified why this is happening.
The issue appears to be related to daemonset Pod deletion, a Pod gets deleted causing `deletePod` to trigger, this lowers the expectations by `del: -1` as expected. 
```
I0328 15:44:49.470363       1 daemon_controller.go:686] Pod fluentd-gv467 deleted.
I0328 15:44:49.470387       1 controller_utils.go:235] Lowered expectations &controller.ControlleeExpectations{add:0, del:-1, key:"kube-system/fluentd", timestamp:time.Time{wall:0xbf1f58d9e8ff2cb3, ext:238457973525, loc:(*time.Location)(0x9578760)}}
```
The delete pod event then triggers a sync of the daemonset, in which it finds `podsToDelete` which involves fetching pods from the informer cache
```go
pods, err := dsc.podLister.Pods(ds.Namespace).List(labels.Everything())
```
This is somehow returning the deleted pod. You can see in the logs here `syncNodes` thinks that it needs to delete the pod that was deleted. It is also setting the expectation back to `del: 1` after the delete pod event previously lowered it by 1.
```
I0328 15:44:49.473164       1 controller_utils.go:218] Setting expectations &controller.ControlleeExpectations{add:0, del:1, key:"kube-system/fluentd", timestamp:time.Time{wall:0xbf1f58dc5c33c66b, ext:248243317929, loc:(*time.Location)(0x9578760)}}
I0328 15:44:49.473197       1 daemon_controller.go:1016] Nodes needing daemon pods for daemon set fluentd: [], creating 0
I0328 15:44:49.473259       1 daemon_controller.go:1102] Pods to delete for daemon set fluentd: [fluentd-gv467], deleting 1
```
At this point you're stuck as the `syncDaemonSet` function can't get past this line
```go
if !dsc.expectations.SatisfiedExpectations(dsKey) {
		// Only update status.
		return dsc.updateDaemonSetStatus(ds, hash)
	}
```
As the expectation can not be satisfied as syncNodes set it back to `del: 1` and there's nothing it to get it back to `del: 0`. 

So I believe this is what's causing us to get less replicas than expected, the `syncDaemonSet` function is stuck due to the expectation so it can't perform a full sync to schedule the new pods. There's also not much to trigger a new sync once the expectation has expired until a new node comes up or one goes down. 

**Environment**:
- Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.1", GitCommit:"eec55b9ba98609a46fee712359c7b5b365bdd920", GitTreeState:"clean", BuildDate:"2018-12-13T10:39:04Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.7", GitCommit:"6f482974b76db3f1e0f5d24605a9d1d38fad9a2b", GitTreeState:"clean", BuildDate:"2019-03-25T02:41:57Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
AWS
- OS (e.g: `cat /etc/os-release`): 
```
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=2023.5.0
VERSION_ID=2023.5.0
BUILD_ID=2019-03-09-0138
PRETTY_NAME="Container Linux by CoreOS 2023.5.0 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
```
- Kernel (e.g. `uname -a`):
`4.19.25-coreos #1 SMP Sat Mar 9 01:05:06 -00 2019 x86_64 Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz GenuineIntel GNU/Linux`

**Comment by Joseph-Irving** [comment]
@kubernetes/sig-apps-bugs I investigated this as much as I could without getting some external input, it feels like it's some kind of race condition in the informer cache. If anyone has ideas of if places to dig deeper that would be handy.

**Comment by Joseph-Irving** [comment]
This problem is definitely related to node termination, it's trying to delete the already deleted pod from the node that's been terminated.

The problem seems to go away if `--feature-gates=ScheduleDaemonSetPods=false` is set, I think although this kind of race condition may have always been possible, it's being frequently caused via 
```go
if utilfeature.DefaultFeatureGate.Enabled(features.ScheduleDaemonSetPods) {
		podsToDelete = append(podsToDelete, getPodsWithoutNode(nodeList, nodeToDaemonPods)...)
	}
```
Which is finding the non-existent pod in the pod cache and seeing that it is assigned to a non-running node, resulting in a request to delete the non-existent pod.

cc @kubernetes/sig-scheduling-bugs

**Comment by bsalamat** [comment]
/assign @krzysztof-jastrzebski 

@krzysztof-jastrzebski could you please take a look at this? It seems that it is related to the code you added in #73401.

**Comment by krzysztof-jastrzebski** [comment]
Thanks for reporting the bug.
Bug was introduced in #73401 which added deleting pods without nodes. There is race condition between removing pod from cache and listing pods for DaemonSet. Bug occurs during following scenario:
1. Node is deleted.
2. Pod for node is deleted by PodGCController.
3. Delete handler is executed by informer (https://github.com/krzysztof-jastrzebski/kubernetes/blob/7186d3ba8ba3df736f6a41e3ab773db3a3698f3f/pkg/controller/daemon/daemon_controller.go#L642)
4. Daemon set is added to queue by delete handler (https://github.com/krzysztof-jastrzebski/kubernetes/blob/7186d3ba8ba3df736f6a41e3ab773db3a3698f3f/pkg/controller/daemon/daemon_controller.go#L685)
5. DaemonSet is processed and pods are taken from cache before deleted pod is removed.
6. Node for deleted pod is not found so we add deleted pod to podsToDelete list (https://github.com/krzysztof-jastrzebski/kubernetes/blob/7186d3ba8ba3df736f6a41e3ab773db3a3698f3f/pkg/controller/daemon/daemon_controller.go#L9680).
7. Expectations are set to del:1(https://github.com/krzysztof-jastrzebski/kubernetes/blob/7186d3ba8ba3df736f6a41e3ab773db3a3698f3f/pkg/controller/daemon/daemon_controller.go#L1002)
8. Deletion succeeds as pod is not found(https://github.com/krzysztof-jastrzebski/kubernetes/blob/7186d3ba8ba3df736f6a41e3ab773db3a3698f3f/pkg/controller/daemon/daemon_controller.go#L1089).
9. Expectations are never met so we need to wait till expectations expire.

As fix I will send a change to #73401 so only pods that can't be deleted by PodGCController are deleted by DaemonSet controller.
